controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java

```diff
@@ -172,52 +166,54 @@ public void triggerReadyJobs(LockedApplication application) {
             if (jobStatus.isRunning(jobTimeoutLimit())) continue;
 
             // Collect the subset of next jobs which have not run with the last changes
-            List<JobType> nextToTrigger = new ArrayList<>();
-            for (JobType nextJobType : order.nextAfter(jobType, application)) {
+            // TODO jvenstad: Change to be step-centric.
+            List<JobType> nextJobs = order.nextAfter(jobType, application);
+            for (JobType nextJobType : nextJobs) {
                 JobStatus nextStatus = application.deploymentJobs().jobStatus().get(nextJobType);
                 if (changesAvailable(application, jobStatus, nextStatus) || nextStatus.isHanging(jobTimeoutLimit()))
-                    nextToTrigger.add(nextJobType);
+                    application = trigger(new Triggering(application, nextJobType, false, "Available change in " + jobType.jobName()), nextJobs, false);
             }
-            // Trigger them in parallel
-            application = trigger(nextToTrigger, application, "Available change in " + jobType.jobName());
             controller.applications().store(application);
         }
     }
 
     /**
-     * Returns true if the previous job has completed successfully with a application version and/or  Vespa version
-     * which is newer (different) than the one last completed successfully in next
+     * Trigger a job for an application, if allowed
+     *
+     * @param triggering the triggering to execute, i.e., application, job type and reason
+     * @param concurrentlyWith production jobs that may run concurrently with the job to trigger
+     * @param force true to disable checks which should normally prevent this triggering from happening
+     * @return the application in the triggered state, if actually triggered. This *must* be stored by the caller
      */
-    private boolean changesAvailable(Application application, JobStatus previous, JobStatus next) {
-        if ( ! application.change().isPresent()) return false;
-        if (next == null) return true;
-
-        if (next.type().isTest()) {
-            // Is it not yet this job's turn to upgrade?
-            if ( ! lastSuccessfulIs(application.change(), previous.type(), application))
-                return false;
-
-            // Has the upgrade test already been done?
-            if (lastSuccessfulIs(application.change(), next.type(), application))
-                return false;
-        }
-        else if (next.type().isProduction()) {
-            // Is the target version tested?
-            if ( ! lastSuccessfulIs(application.change(), JobType.stagingTest, application))
-                return false;
+    // TODO jvenstad: Replace with (Collection<JobType> concurrentlyWith) to allow, e.g., concurrent retries.
+    public LockedApplication trigger(Triggering triggering, Collection<JobType> concurrentlyWith, boolean force) {
```

Comment:
This is by no means final. However, I need this class for when jobs to run are computed on-demand, and sorted (these will be sorted), so I thought I could as well get it in there sooner, rather than later.

Reply:
The final pseudo will be something like ```java
List<Triggering> jobs = findAllJobsForAllApplicationsThatMayRunNow(jobType);
jobs.sort();
trigger(jobType.isCapacityConstrained() ? jobs.subList(0, 1) : jobs);
```

Reply:
OK. Merge at will 👍

Reply:
Thanks :)